pennylane/devices/qubit_mixed/measure.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code relevant for performing measurements on a qutrit mixed state.
+"""
+
+from collections.abc import Callable
+from string import ascii_letters as alphabet
+
+from pennylane import math, queuing
+from pennylane.measurements import (
+    ExpectationMP,
+    MeasurementProcess,
+    ProbabilityMP,
+    StateMeasurement,
+    StateMP,
+    VarianceMP,
+)
+from pennylane.ops import Sum
+from pennylane.typing import TensorLike
+from pennylane.wires import Wires
+
+from .apply_operation import _get_num_wires, apply_operation
+
+
+def _reshape_state_as_matrix(state, num_wires):
+    """Given a non-flat, potentially batched state, flatten it to square matrix or matrices if batched.
+
+    Args:
+        state (TensorLike): A state that needs to be reshaped to a square matrix or matrices if batched
+        num_wires (int): The number of wires the state represents
+
+    Returns:
+        Tensorlike: A reshaped, square state, with an extra batch dimension if necessary
+    """
+    dim = 2**num_wires
+    batch_size = math.get_batch_size(state, ((2,) * (num_wires * 2)), dim**2)
+    shape = (batch_size, dim, dim) if batch_size is not None else (dim, dim)
+    return math.reshape(state, shape)
+
+
+def calculate_expval(
+    measurementprocess: ExpectationMP,
+    state: TensorLike,
+    is_state_batched: bool = False,
+    readout_errors: list[Callable] = None,
+) -> TensorLike:
+    """Measure the expectation value of an observable.
+
+    Args:
+        measurementprocess (ExpectationMP): measurement process to apply to the state.
+        state (TensorLike): the state to measure.
+        is_state_batched (bool): whether the state is batched or not.
+        readout_errors (List[Callable]): List of chanels to apply to each wire being measured
+        to simulate readout errors.
+
+    Returns:
+        TensorLike: expectation value of observable wrt the state.
+    """
+    probs = calculate_probability(measurementprocess, state, is_state_batched, readout_errors)
```

Comment:
I thought the different measurement processes already have `process_density_matrix`? Are we not able to use those?

Reply:
Oh that's so true. I forgot the truth that they were created by me myself 😅 
The reason why they are like these is that to save time I did minimal modification based on the `qutrit_mixed`; when they were implemented, it was an age there is no `process_density_matrix`. 

How about this: we still leave them as current version because things are running OKish without bugs. Later on after the Epic concludes I initiate two other stories, one for this measure and the other for qutrit mixed.

Reply:
I feel that if we put this off now, realistically it will never be worked on. If it doesn't require large architectural change to what we currently have, I'd prefer we do this properly now.

Reply:
[sc-79423]